File: frontend/src/types/replacement.ts
import dayjs from "dayjs";
import { BreachT } from "./breach";

export type ReplacementCategoryT = "can_do" | "could_do" | "cant_do";

export type FilterHitsT = {
  isntFilteredOut: boolean;
  filterLabels: string[];
};

export type OverlapHitsT = {
  hasntOverlap: boolean;
  overlapAssignmentIds: string[];
};

export type ConstraintHitsT = {
  meetsConstraints: boolean;
  breaches: BreachT[];
};

export type RequestHitsT = {
  hasNoRequestConflict: boolean;
  conflictingRequestIds: string[];
};

export type MonthlyDutiesImplicationsT = {
  newNumberMonthlyDuties: number;
  newMonthlyDutiesDelta: number;
  meetsTarget: boolean;
};

export type WeeklyWorkTimeImplicationsT = {
  newWeeklyWorkedMinutes: number;
  newWeeklyTimeDeltaMinutes: number;
  meetsTarget: boolean;
};

export type LTMIndicatorT = {
  count: number;
  lastDate: dayjs.Dayjs | null;
};

export type ReplacementImplicationsT = {
  // Can't do (hard constraints)
  isEmployed: boolean;
  hasSpecialty: boolean;
  isntOnLeave: boolean;
  filterHits: FilterHitsT;
  overlapHits: OverlapHitsT;
  hardConstraintHits: ConstraintHitsT;
  requestHits: RequestHitsT;

  // Could do (soft constraints)
  softConstraintHits: ConstraintHitsT;
  newMonthlyDuties: MonthlyDutiesImplicationsT;
  newWeeklyTime: WeeklyWorkTimeImplicationsT;

  // Indicators (informational)
  nbTimesDidShiftLtm: LTMIndicatorT;
  nbTimesWorkedWeekdayLtm: LTMIndicatorT;
};

export type ReplacementCandidateT = {
  workerId: string;
  workerName: string;
  rank: number;
  replacementCategory: ReplacementCategoryT;
  replacementImplications: ReplacementImplicationsT;
  mostConstrainingReason: string;
};

/**
 * Convert raw API data to ReplacementCandidateT
 */
export function toReplacementCandidateT(data: any): ReplacementCandidateT {
  return {
    ...data,
    replacementImplications: {
      ...data.replacementImplications,
      nbTimesDidShiftLtm: {
        count: data.replacementImplications.nbTimesDidShiftLtm.count,
        lastDate: data.replacementImplications.nbTimesDidShiftLtm.lastDate
          ? dayjs.unix(data.replacementImplications.nbTimesDidShiftLtm.lastDate)
          : null,
      },
      nbTimesWorkedWeekdayLtm: {
        count: data.replacementImplications.nbTimesWorkedWeekdayLtm.count,
        lastDate: data.replacementImplications.nbTimesWorkedWeekdayLtm.lastDate
          ? dayjs.unix(
              data.replacementImplications.nbTimesWorkedWeekdayLtm.lastDate,
            )
          : null,
      },
    },
  };
}

File: backend/shared/src/shared/schemas/dto/replacement.py
from typing import List

from pydantic import BaseModel

from shared.schemas.dto.breach import BreachDTO


class FilterHitsDTO(BaseModel):
    isntFilteredOut: bool
    filterLabels: List[str]


class OverlapHitsDTO(BaseModel):
    hasntOverlap: bool
    overlapAssignmentIds: List[str]


class ConstraintHitsDTO(BaseModel):
    meetsConstraints: bool
    breaches: List[BreachDTO]


class RequestHitsDTO(BaseModel):
    hasNoRequestConflict: bool
    conflictingRequestIds: List[str]


class MonthlyDutiesImplicationsDTO(BaseModel):
    newNumberMonthlyDuties: int
    newMonthlyDutiesDelta: int
    meetsTarget: bool


class WeeklyWorkTimeImplicationsDTO(BaseModel):
    newWeeklyWorkedMinutes: int
    newWeeklyTimeDeltaMinutes: int
    meetsTarget: bool


class LTMIndicatorDTO(BaseModel):
    count: int
    lastDate: float | None  # Unix timestamp


class ReplacementImplicationsDTO(BaseModel):
    # Can't do (hard constraints)
    isEmployed: bool
    hasSpecialty: bool
    isntOnLeave: bool
    filterHits: FilterHitsDTO
    overlapHits: OverlapHitsDTO
    hardConstraintHits: ConstraintHitsDTO
    requestHits: RequestHitsDTO

    # Could do (soft constraints)
    softConstraintHits: ConstraintHitsDTO
    newMonthlyDuties: MonthlyDutiesImplicationsDTO
    newWeeklyTime: WeeklyWorkTimeImplicationsDTO

    # Indicators (informational)
    nbTimesDidShiftLtm: LTMIndicatorDTO
    nbTimesWorkedWeekdayLtm: LTMIndicatorDTO


class ReplacementCandidateDTO(BaseModel):
    workerId: str
    workerName: str
    rank: int
    replacementCategory: str  # "can_do", "could_do", "cant_do"
    replacementImplications: ReplacementImplicationsDTO
    mostConstrainingReason: str


class AssignmentImplicationDTO(BaseModel):
    assignmentId: str
    implications: ReplacementImplicationsDTO
    replacementCategory: str
    mostConstrainingReason: str


class SwapAssignmentInfoDTO(BaseModel):
    workerId: str
    workerName: str
    preSwap: List[AssignmentImplicationDTO]
    postSwap: List[AssignmentImplicationDTO]


class SwapValidationResultDTO(BaseModel):
    isValid: bool
    workerAInfo: SwapAssignmentInfoDTO
    workerBInfo: SwapAssignmentInfoDTO
    validationKey: str
